salt/modules/x509.py

```diff
@@ -680,7 +697,24 @@ def write_pem(text, path, pem_type=None):
     '''
     old_umask = os.umask(0o77)
     text = get_pem_entry(text, pem_type=pem_type)
-    salt.utils.fopen(path, 'w').write(text)
+    _dhparams = ''
+    _private_key = ''
+    if pem_type and pem_type == 'CERTIFICATE' and os.path.isfile(path):
+        _filecontents = _text_or_file(path)
+        try:
+            _dhparams = get_pem_entry(_filecontents, 'DH PARAMETERS')
+        except salt.exceptions.SaltInvocationError:
+            pass
+        try:
+            _private_key = get_pem_entry(_filecontents, 'RSA PRIVATE KEY')
+        except salt.exceptions.SaltInvocationError:
+            pass
+    with salt.utils.fopen(path, 'w') as _fp:
+        if pem_type and pem_type == 'CERTIFICATE' and _private_key:
```

Comment:
This is going to write out the private key even if the requested type to write out was CERTIFICATE? I don't think that's a safe assumption to make.

Perhaps pem_type could take multiple values so you could pass in both CERTIFICATE and RSA PRIVATE KEY, but if only certificate is specified, I think only certificate should be written.